### ISSUE TITLE
Persist session keys

### ISSUE THREAD

**Issue by picklenerd** [boltlabs-inc/key-mgmt#213]
Session keys need to be persisted on the server so that they can be used for authenticated operations.

In the short term, we can store them in a map within the `LockKeeperKeyServer` struct along with a timestamp that can be used to invalidate the session key. For the long term, we should consider something like Redis or another alternative that's quick and can manage invalidation automatically.

The interface to retrieve a session key from storage should return an error if the session key is expired and delete the stored key.

The index for the map can be the `UserId` or `AccountName`. The `UserId` is probably the most appropriate but it would need to be pulled from the database for each request since we only know the `AccountName` from the request's headers. We can discuss which is better to use once you get to this point.

The session key should be available in the OPAQUE server finish result that's generated in the authenticate operation. Figuring out how to get this back to the server state might be tricky. The `Operation` trait sends a mutable `Context` which could be used for this, but there might be a better solution that requires a little bit of refactoring. We can discuss this as well.

The session key is already stored on the client side so the client code should be a useful reference for how the server might work.

**Comment by marsella** [comment]
To clarify, this issue is about persisting the session key but not necessarily _using_ it, right? E.g. I would not expect this issue to update the message conversion and proto format to encrypt messages under the session key (and decrypt+validate received messages). I would not expect it to [resolve this TODO](https://github.com/boltlabs-inc/key-mgmt/blob/develop/lock-keeper/src/types/authenticate.rs#L48-L51):
```rust
   // TODO #186: This struct should be authenticated! Update message conversion to
    // authenticate on serialization and check authentication on
    // deserialization.
    impl_message_conversion!(SendUserId);
```

**Comment by picklenerd** [comment]
Correct. This issue is just to store the session key on the server so that it's available later when we need it. The client already stores it so the feature feels half-finished.

**Comment by gatoWololo** [comment]
I have looked through this issue and the code and I think I have a pretty good sense of all the relevant parts.

> we can store them in a map within the LockKeeperKeyServer struct along with a timestamp that can be used to invalidate the session key

This part makes sense.

> The session key is already stored on the client side so the client code should be a useful reference for how the server might work.

I see how the client does it and have modified the code so the session key (`ServerLoginFinishResult<OpaqueCipherSuite>`) is returned all the way up to the `operation` method for the `impl Operation for Authenticate`. Like the original issue message stated, the question is how to get it back to the `LockKeeperKeyServer` from here. You mentioned we could use the mutable `Context` but that doesn't seem ideal.

The alternative I am thinking is modifying the `Operation` trait, so instead of the signature for `operation` looking like:
```rust
pub(crate) trait Operation: Sized + Send + 'static {
    /// Core logic for a given operation.
    async fn operation(
        self,
        channel: &mut ServerChannel,
        context: &mut Context,
    ) -> Result<(), LockKeeperServerError>;
...
}
```
We define an associated type and make that the return output of `operation`:
```rust
pub(crate) trait Operation: Sized + Send + 'static {
    type OperationReturn;

    /// Core logic for a given operation.
    async fn operation(
        self,
        channel: &mut ServerChannel,
        context: &mut Context,
    ) -> Result<OperationReturn, LockKeeperServerError>;
...
}
```
From there, I am still unclear how the signature to `Operation::handle_request` should be modified to pass this value back...

The other important piece I have been thinking about is going through and cleaning up all the expired session keys. It is easy to clean them up if they are referenced and we notice they are expired, but if a session key is never again referenced, it will never be cleaned up. Is this something we want to handle as part of this issue?

I could imagine naive schemes that iterate through all session keys on any query into our map and clean up expired keys. I can also imagine better schemes that don't require a linear scan every time we fetch a session key!

**Comment by picklenerd** [comment]
We discussed this over a call. Here's a summary:

We decided not to change the return type of `operation` since this is likely the only situation that would use that feature. We're going to add the session storage to the `Context` wrapped in an `Arc<RwLock>`. The authenticate operation will insert the session key into storage  and other operations will read from it to get the session key. We could also add a method to the `Context` to get the session key for the current user to make things easier.

We decided not to manually clean up old sessions right now. They will still be removed if the user tries to call an authenticated operation with a stale session, but we won't have any kind of automated cleanup. The thinking here is that this is a short term solution to storage so that we can implement authentication properly. The longer term solution will likely involve `redis` or some other self-evicting cache option. We don't currently have any long-running server processes and the cache will be dropped any time the server is shut down or restarted so this shouldn't cause any issues in the short term. This will need to be addressed once we deploy this to a real server.

We also discussed whether session storage should be directly stored as a `HashMap` or wrapped in an abstraction. We decided to go with an abstraction so that we can provide convenience methods and change the underlying implementation without breaking too much code.

**Comment by gatoWololo** [comment]
Closed by #301
